src/platform/amd/acp_7_0/lib/clk.c

```diff
@@ -0,0 +1,491 @@
+	while (delay_cnt > 0) {
+		acp_srbm_cycle_sts = (acp_srbm_cycle_sts_t)io_reg_read(PU_REGISTER_BASE +
+								ACP_SRBM_CYCLE_STS);
+		if (!acp_srbm_cycle_sts.bits.srbm_clients_sts)
+			return;
+		delay_cnt--;
+	}
+}
+
+void acp_7_0_reg_wait(void)
+{
+	int test_count = 0;
+	int val = 0;
+
+	for (test_count = 0; test_count < 255; test_count++) {
+		val = acp_reg_read_via_smn(CLK7_CLK_FSM_STATUS, sizeof(int));
+		val = val * 0;
+	}
+}
+
+void acp_7_0_get_boot_ref_clock(float *boot_ref_clk)
+{
+	clk7_clk_pll_refclk_startup_t boot_ref_clk_startup;
+	clk7_clk_pll_req_u_t clk7_clk_pll_req;
+	clk7_spll_field_2_t clk7_spll_field;
+	uint32_t spinediv      = 1;
+	float fract_part    = 0.0f;
+	float final_refclk  = 0.0f;
+
+	boot_ref_clk_startup.u32all =
+		acp_reg_read_via_smn(CLK7_CLK_PLL_REFCLK_RATE_STARTUP, sizeof(int));
+
+	clk7_spll_field_9_t  clk_spll_field_9;
+
+	clk_spll_field_9.u32all = 0x0;
+	clk_spll_field_9.u32all = acp_reg_read_via_smn(CLK7_SPLL_FIELD_9, sizeof(int));
+
+	if (clk_spll_field_9.bitfields.spll_dpll_cfg_3 == 0x2)
+		final_refclk = ((32768.0f * 128.0f) / 1000000.0f);
+	else
+		final_refclk = (float)boot_ref_clk_startup.bitfields.main_pll_ref_clk_rate_startup;
+
+	clk7_clk_pll_req.u32all = acp_reg_read_via_smn(CLK7_CLK_PLL_REQ, sizeof(int));
+	clk7_spll_field.u32all = acp_reg_read_via_smn(CLK7_SPLL_FIELD_2, sizeof(int));
+
+	spinediv = (1 << clk7_clk_pll_req.bitfields.pllspinediv);
+
+	if (clk7_spll_field.bitfields.spll_fracn_en == 1)
+		fract_part = (float)(clk7_clk_pll_req.bitfields.fbmult_frac / (float)65536.0f);
+
+	*boot_ref_clk = (float)(((final_refclk) * (clk7_clk_pll_req.bitfields.fbmult_int +
+					fract_part)) / (float)spinediv);
+}
+
+void acp_change_clock_notify(uint32_t clock_freq)
+{
+	volatile clk7_clk1_dfs_cntl_u_t dfs_cntl;
+	volatile clk7_clk1_bypass_cntl_u_t bypass_cntl;
+	volatile clk7_clk1_dfs_status_u_t dfs_status;
+	volatile uint32_t updated_clk = 0;
+	float did = 0.0f;
+	float fraction_val = 0.0f;
+	uint32_t int_did_val = 0;
+	float boot_ref_clk;
+	acp_clock_type_t clock_type = acp_aclk_clock;
+
+	acp_7_0_get_boot_ref_clock(&boot_ref_clk);
+
+	tr_info(&acp_clk_tr, "acp_change_clock_notify clock_freq : %d clock_type : %d",
+		clock_freq, clock_type);
+
+	fraction_val = (float)(clock_freq / (float)1000000.0f);
+	clock_freq   = (clock_freq / 1000000);
+	if (acp_aclk_clock == clock_type) {
+		bypass_cntl.u32all = acp_reg_read_via_smn(CLK7_CLK1_BYPASS_CNTL, sizeof(int));
+		dfs_cntl.u32all = acp_reg_read_via_smn(CLK7_CLK1_DFS_CNTL, sizeof(int));
+	} else if (acp_sclk_clock == clock_type) {
+		bypass_cntl.u32all = acp_reg_read_via_smn(CLK7_CLK0_BYPASS_CNTL, sizeof(int));
+		dfs_cntl.u32all = acp_reg_read_via_smn(CLK7_CLK0_DFS_CNTL, sizeof(int));
+	}
+
+	bypass_cntl.bitfields.CLK1_BYPASS_DIV = 0;
+
+	if (clock_freq == 6 || clock_freq == 0) {
+		did = 128;
+		dfs_cntl.bitfields.CLK1_DIVIDER = 0x7F;
+		bypass_cntl.bitfields.CLK1_BYPASS_DIV = 0xF;
+	} else {
+		did  = (float)(boot_ref_clk / (float)fraction_val);
```

Comment:
all variables are already of type `float`, no need for any type-casts

Reply:
Done